returnn/config.py

```diff
@@ -366,6 +323,93 @@ def _sis_hash(self):
+    """
+    Returnn config class with V2 serialization.
+
+    Can be constructed like the classic `ReturnnConfig` or from an existing config like this:
+    ```python
+    from i6_core.returnn import ReturnnConfig, ReturnnConfigV2, ReturnnTrainingJob
+
+    config = ReturnnConfig(...)  # your usual config
+    config_v2_serialization = ReturnnConfigV2({})
+    config_v2_serialization.update(config)
+    train_job = ReturnnTrainingJob(config=config_v2_serialization)
+    # train job will use V2 serialization now
+    ```
+
+    During serialization, we call `instanciate_delayed_copy` on the config dict.
+    This will resolve all delayed values and path objects into their actual values/str representation.
+    However, this will only catch those values that are reachable for the `tree` library,
+    i.e. all values contained in (nested) dictionaries/sequences.
+
+    If you have delayed values/path objects which are not directly reachable from the config dict,
+    e.g. by being inside some custom object instead of a dict/list/tuple/set,
+    then you need to resolve them manually before serialization or special-case their
+    `__reduce__` behavior via the :func:`in_serialize_config` flag.
+
+    See also :func:`i6_core.serialization.serialization_v2.serialize_config` and the corresponding module.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        assert not self.staged_network_dict, "V2 serialization does not support staged net dicts"
+
+    def _serialize(self) -> str:
+        import tree
+
+        from i6_core.serialization import ExternalImport
+        from i6_core.serialization.serialization_v2 import get_base_sys_path_list, serialize_config
+
+        assert not self.staged_network_dict, "V2 serialization does not support staged net dicts"
+
+        self.check_consistency()
+
+        config = instanciate_delayed_copy(self.config)
+        post_config = instanciate_delayed_copy(self.post_config)
+
+        # I'm not really sure about it.
+        # Our automatic mechanism will find direct imports (e.g. i6_experiments).
+        # However, it will not find indirect imports (e.g. sisyphus),
+        # and thus the generated code might fail.
+        # So add all other paths here which we currently have.
+        # (While testing this, this was sisyphus + returnn + recipes,
+        #  but returnn is excluded below.)
+        sys_paths = set(get_base_sys_path_list())
+        extra_sys_paths = [p for p in sys.path if p not in sys_paths]
+
+        # Handle ExternalImports
+        extra_sys_paths += [
+            item.import_path
+            for item in tree.flatten(self.python_prolog) + tree.flatten(self.python_epilog)
+            if isinstance(item, ExternalImport)
+        ]
+
+        python_prolog_code = unparse_python(self.python_prolog, hash_full_python_code=True)
+        python_epilog_code = unparse_python(self.python_epilog, hash_full_python_code=True)
```

Comment:
why is `hash_full_python_code` hardcoded here? It seems the only thing it does is prevent us from using `DelayedBase` objects in prolog/epilog without actually affecting the hash

```suggestion
        python_prolog_code = unparse_python(self.python_prolog, hash_full_python_code=self.hash_full_python_code)
        python_epilog_code = unparse_python(self.python_epilog, hash_full_python_code=self.hash_full_python_code)
```

of course, if the parameter does not make sense in the new implementation, then I would rather add an
```
assert hash_full_python_code == True
```
to the `__init__` of `ReturnnConfigV2`.

But then I am wondering that hashing and serialization should be independent of each other.. so what is the case here?

EDIT: I read below
> There is a :class:`SerializerObject` base class, deriving from Sisyphus :class:`DelayedBase`.

so actually the `# Handle ExternalImports` part does not make sense if we disallow them in the prolog/epilog.